In [239]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.preprocessing import MinMaxScaler, OrdinalEncoder, OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score
import warnings
warnings.filterwarnings("ignore")
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier


In [16]:
df_base=pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")

In [17]:
df_base.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [18]:
df_base.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [19]:
df_base.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.123129  
std                   3.222430              3.568136  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

In [20]:
df_base.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [21]:
df_copy=df_base.copy()

In [22]:
missing_percentages = (df_copy.isnull().sum() / len(df_copy)) * 100
missing_data = pd.DataFrame({'Column': missing_percentages.index, 'MissingPercentage': missing_percentages.values})
missing_data.sort_values(by='MissingPercentage', ascending=False)

Column  MissingPercentage
0                        Age                0.0
26             StandardHours                0.0
20        NumCompaniesWorked                0.0
21                    Over18                0.0
22                  OverTime                0.0
23         PercentSalaryHike                0.0
24         PerformanceRating                0.0
25  RelationshipSatisfaction                0.0
27          StockOptionLevel                0.0
18             MonthlyIncome                0.0
28         TotalWorkingYears                0.0
29     TrainingTimesLastYear                0.0
30           WorkLifeBalance                0.0
31            YearsAtCompany                0.0
32        YearsInCurrentRole                0.0
33   YearsSinceLastPromotion                0.0
19               MonthlyRate                0.0
17             MaritalStatus                0.0
1                  Attrition                0.0
8              EmployeeCount                0.0
2             BusinessTravel                0.0
3                  DailyRate                0.0
4                 Department                0.0
5           DistanceFromHome                0.0
6                  Education                0.0
7             EducationField                0.0
9             EmployeeNumber                0.0
16           JobSatisfaction                0.0
10   EnvironmentSatisfaction                0.0
11                    Gender                0.0
12                HourlyRate                0.0
13            JobInvolvement                0.0
14                  JobLevel                0.0
15                   JobRole                0.0
34      YearsWithCurrManager                0.0

In [23]:
df_copy.nunique()
#employee_number(high cardinality)
#Employeecount (only one value)
#Over18 (only one value)
#StandardHours (only one value)

Age                           43
Attrition                      2
BusinessTravel                 3
DailyRate                    886
Department                     3
DistanceFromHome              29
Education                      5
EducationField                 6
EmployeeCount                  1
EmployeeNumber              1470
EnvironmentSatisfaction        4
Gender                         2
HourlyRate                    71
JobInvolvement                 4
JobLevel                       5
JobRole                        9
JobSatisfaction                4
MaritalStatus                  3
MonthlyIncome               1349
MonthlyRate                 1427
NumCompaniesWorked            10
Over18                         1
OverTime                       2
PercentSalaryHike             15
PerformanceRating              2
RelationshipSatisfaction       4
StandardHours                  1
StockOptionLevel               4
TotalWorkingYears             40
TrainingTimesLastYear          7
WorkLifeBa

In [24]:
df_copy.drop(columns=['EmployeeCount', 'Over18', 'StandardHours','EmployeeNumber'], inplace=True)

In [56]:
attr=df_copy.copy()

In [177]:
attr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 31 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EnvironmentSatisfaction   1470 non-null   int64 
 9   Gender                    1470 non-null   object
 10  HourlyRate                1470 non-null   int64 
 11  JobInvolvement            1470 non-null   int64 
 12  JobLevel                  1470 non-null   int64 
 13  JobRole                   1470 non-null   object
 14  JobSatisfaction         

In [57]:
#splitting
X = attr.drop('Attrition', axis = 1)
y = attr.Attrition

In [58]:
#splitting data into train and test
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.3)
print(X_train.shape) ; print(X_test.shape)

(1029, 30)
(441, 30)


In [59]:
#OrdinalEncoder 'BusinessTravel', 'Education', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobSatisfaction','WorkLifeBalance',
#'PerformanceRating','RelationshipSatisfaction'
#binary gender  OverTime
#onehotencoder 'Department', 'EducationField', 'JobRole', 'MaritalStatus','BusinessTravel


In [165]:
categorical_feature_ordinal= attr[[ 'Education', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobSatisfaction','WorkLifeBalance',
'PerformanceRating','RelationshipSatisfaction','JobLevel','StockOptionLevel']]
categorical_feature_ordinal.head()


Education  EnvironmentSatisfaction  JobInvolvement  JobSatisfaction  \
0          2                        2               3                4   
1          1                        3               2                2   
2          2                        4               2                3   
3          4                        4               3                3   
4          1                        1               3                2   

   WorkLifeBalance  PerformanceRating  RelationshipSatisfaction  JobLevel  \
0                1                  3                         1         2   
1                3                  4                         4         2   
2                3                  3                         2         1   
3                3                  3                         3         1   
4                3                  3                         4         1   

   StockOptionLevel  
0                 0  
1                 1  
2                 0  
3                 0  
4                 1

In [178]:
categorical_feature=['Department', 'EducationField', 'JobRole', 'MaritalStatus','BusinessTravel']

In [179]:
numerical_feature= ['Age','YearsWithCurrManager','YearsSinceLastPromotion','YearsInCurrentRole','YearsAtCompany','TrainingTimesLastYear','TotalWorkingYears',
                          'PercentSalaryHike','NumCompaniesWorked','MonthlyIncome','MonthlyRate','HourlyRate','DistanceFromHome','DailyRate']

In [180]:
binary_features = ['Gender','OverTime']

In [181]:
from sklearn.preprocessing import FunctionTransformer
def binary_transformer(X):
    return X.apply(lambda x: x.map({'Male': 0, 'Female': 1, 'Yes': 1, 'No': 0}))
binary_preprocessor = FunctionTransformer(binary_transformer)

In [182]:
categorical_transformer = Pipeline(steps=[
    ('onehot', OneHotEncoder(handle_unknown='ignore'))]) # One hot encoding

In [183]:
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())])

In [184]:
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numerical_feature),
        ('cat', categorical_transformer, categorical_feature),
        ('binary', binary_preprocessor, binary_features)], 
        remainder = 'passthrough')

In [186]:
y_train.replace({'No':0,
                'Yes':1}, inplace = True)
y_test.replace({'No':0,
                'Yes':1}, inplace = True)
print(y_train.value_counts())
print(y_test.value_counts())

Attrition
0    876
1    153
Name: count, dtype: int64
Attrition
0    357
1     84
Name: count, dtype: int64


In [240]:
models = {
    'random_forest': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', RandomForestClassifier())
    ]),
    'logistic_regression': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', LogisticRegression())
    ]),
    'decision_tree': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', DecisionTreeClassifier())
    ]),
    'gradient_boosting': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', GradientBoostingClassifier())
    ]),
    'xgboost': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', XGBClassifier(use_label_encoder=False, eval_metric='mlogloss'))
    ]),
    'adaboost': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', AdaBoostClassifier())
    ])
}

# Fit models
for name, pipeline in models.items():
    pipeline.fit(X_train, y_train)
   



In [228]:
def predict(model):
    y_predict = model.predict(X_test)
    print('Accuracy: {:.2f}%'.format(accuracy_score(y_test, y_predict) * 100))
    print('Precision: {:.2f}%'.format(precision_score(y_test, y_predict, average='binary') * 100))
    print('Recall: {:.2f}%'.format(recall_score(y_test, y_predict, average='binary') * 100))
    print('F1_Score: {:.2f}%'.format(f1_score(y_test, y_predict, average='binary') * 100))

    # Generate the confusion matrix
    cm = confusion_matrix(y_test, y_predict)

    # Print the confusion matrix
    print("Confusion Matrix:")
    print("          Predicted:")
    print("             Neg    Pos")
    print("Actual Neg  {0: <6} {1: <6}".format(cm[0, 0], cm[0, 1]))
    print("       Pos  {0: <6} {1: <6}".format(cm[1, 0], cm[1, 1]))

In [229]:
predict(models['random_forest'])

Accuracy: 84.58%
Precision: 86.36%
Recall: 22.62%
F1_Score: 35.85%
Confusion Matrix:
          Predicted:
             Neg    Pos
Actual Neg  354    3     
       Pos  65     19    


In [232]:
predict(models['logistic_regression'])

Accuracy: 85.71%
Precision: 71.43%
Recall: 41.67%
F1_Score: 52.63%
Confusion Matrix:
          Predicted:
             Neg    Pos
Actual Neg  343    14    
       Pos  49     35    


In [234]:
predict(models['decision_tree'])

Accuracy: 78.91%
Precision: 44.30%
Recall: 41.67%
F1_Score: 42.94%
Confusion Matrix:
          Predicted:
             Neg    Pos
Actual Neg  313    44    
       Pos  49     35    


In [241]:
predict(models['gradient_boosting'])

Accuracy: 82.77%
Precision: 61.11%
Recall: 26.19%
F1_Score: 36.67%
Confusion Matrix:
          Predicted:
             Neg    Pos
Actual Neg  343    14    
       Pos  62     22    


In [242]:
predict(models['xgboost'])

Accuracy: 82.99%
Precision: 60.47%
Recall: 30.95%
F1_Score: 40.94%
Confusion Matrix:
          Predicted:
             Neg    Pos
Actual Neg  340    17    
       Pos  58     26    


In [243]:
predict(models['adaboost'])

Accuracy: 85.49%
Precision: 71.74%
Recall: 39.29%
F1_Score: 50.77%
Confusion Matrix:
          Predicted:
             Neg    Pos
Actual Neg  344    13    
       Pos  51     33    


Decision Tree has the highest Recall correctly predicting 35 out of 84 employees who left 41.67%


Regression had a better accuracy score 52.63% making more correct predictions for both classes than any other model.

#Considering the main goal to identify employees that are more likely to leave the company, the recall score is the one that I'm focusing on here. I don't believe that false positives would be expensive to deal with in this case, so I want the maximum amount of true positives as possible.



Random Forest: A robust ensemble of decision trees, good for handling non-linear data with built-in handling of feature importance.

Logistic Regression: A baseline for binary classification problems

.
Decision Tree: Provides a basis for understanding feature selection and decision-making criteri

a.
Gradient Boosting: Builds trees one at a time, where each new tree helps to correct errors made by the previously trained tr

ee.
XGBoost: An optimized gradient boosting library that is highly efficient, flexible, and porta

ble.
AdaBoost: A boosting algorithm that combines multiple weak classifiers to form a strong classifier.